Trait review

rusty1968 opened the pull request.

rusty1968 committed:
Algorithmic crates.

rusty1968 committed:
sample - how-to write generic drivers

rusty1968 committed:
Addressing review comments:

- Add Message to ECDSA Types.
- New trait Serde to allow for types to be converted from/to its byte
  representation.
- New trait OutputSize to describe a fixed OutputSize.

rusty1968 committed:
Rename package

rusty1968 committed:
Remove trait bound from EcdsaTypes::Curve.

rusty1968 committed:
Serde module
Refactor crypto traits.

rusty1968 committed:
Make it a cargo workspace

rusty1968 committed:
Update workspace Cargo.toml 

Use resolver=2

rusty1968 committed:
Change the definition of Ecdsa traits

rusty1968 committed:
Async version of Digest

rusty1968 committed:
Make Serde a trait bound

rusty1968 committed:
Modified original definition of the Mac crypto function to be able to describe
the Caliptra hardware.

- Remove set_key method from Mac trait
- Separate Mac control and data path.
- Isolate the reset function in a separate trait.
- Solved lifetime issues in init function.

rusty1968 committed:
Cleanup demo code

rusty1968 committed:
Update Digest trait definitions.

rusty1968 committed:
Symmetric cipher trait.

cfrantz commented on src/serde.rs:
```diff
@@ -24,15 +23,15 @@ pub trait ErrorType {
-    fn to_le_bytes(&self, dest: &mut [u8]) -> Result<(),Self::Error>;
-    fn from_bytes_le(bytes: &[u8]) -> Result <Self::OutputType, Self::Error>;
+    fn to_le_bytes(&self, dest: &mut [u8]) -> Result<(), Self::Error>;
+    fn from_bytes_le(bytes: &[u8]) -> Result<Self::OutputType, Self::Error>;
```
Would prefer the ordering of the name components to be consistent:
```
to_le_bytes
from_le_bytes
```

cfrantz commented on src/serde.rs:
```diff
@@ -24,15 +23,15 @@ pub trait ErrorType {
 }
 
-pub trait Serde : ErrorType {
+pub trait Serde: ErrorType {
```
In this context `Serde` refers to serialization and deserialization generally, and not to the well-known crates.io `serde` crate.  Maybe we should choose a different name.

cfrantz commented on proposed-traits/src/digest.rs:
```diff
@@ -65,66 +64,35 @@ pub trait ErrorType {
-pub trait Digest: ErrorType + OutputSize {
-    type InitParams<'a>; // use GAT here to allow for the trait user to add a reference to a peripheral with a specified lifetime.
+pub trait DigestCtrl: ErrorType {
+    type InitParams<'a>: where Self: 'a;
```
Can you show an example demonstrating why InitParams needs to be a GAT?

cfrantz commented on proposed-traits/src/digest.rs:
```diff
@@ -153,5 +115,6 @@ pub trait AsyncDigest: ErrorType + OutputSize {
     ///
     /// A `Result` indicating success or failure. On success, returns `Ok(())`. On failure, returns a `CryptoError`.    
-    fn finalize(&mut self, out: &mut [u8]) -> Result<(), Self::Error>;
+    fn finalize(&mut self, output: &mut [u8]) -> Result<(), Self::Error>;
```
TBD: should finalize return a fixed size item relating to the particular digest implementation, or should it take a mutable slice where it would write the output?

```
pub trait DigestOp: ErrorType {
  type Output;

  // Take a mutable reference to the output type
  fn finalize(&mut self, output: &mut Self::Output) -> Result<(), Self::Error>;

  // or, return the output type:
  fn finalize(&mut self) -> Result<Self::Output, Self::Error>;
}

cfrantz replied:
Also, should `finalize` consume `self`?

cfrantz commented on proposed-traits/src/symm_cipher.rs:
```diff
@@ -0,0 +1,85 @@
+}
+
+
+pub trait SymmetricCipherOp {
```
Do the different cipher modes have an effect on how the SymmetricCipherOp trait should operate?

cfrantz commented on proposed-traits/src/mac.rs:
```diff
@@ -0,0 +1,122 @@
+    /// # Returns
+    ///
+    /// A `Result` indicating success or failure. On success, returns `Ok(())`. On failure, returns a `CryptoError`.    
+    fn finalize(&mut self, tag: &mut [u8]) -> Result<(), Self::Error>;
```
Since finalize is the last operation that you can do, should `finalize` consume `self`.

rusty1968 committed:
Rename Serde trait.

rusty1968 committed:
- Rename DigestCtrl trait to DigestInit
- Do not use GAT to describe the InitParams associated type  of the DigestInit trait.

cfrantz commented on proposed-traits/src/common.rs:
```diff
@@ -0,0 +1,37 @@
+    const OUTPUT_SIZE: usize;
+}
+
+pub trait SerializeDeserialize: ErrorType {
```
See `byteorder` for operating on both endians:
https://docs.rs/byteorder/1.5.0/byteorder/

rusty1968 committed:
- Remove OutputSize trait.
- Rename MacCtrl to MacInit.

rusty1968 committed:
Simulated implementation of the Digest trait.

rusty1968 committed:
Update RsaVerify trait to take &mut self.

rusty1968 committed:
Clock and Reset control.

rusty1968 committed:
Introduce the I2CTarget trait.

rusty1968 committed:
Introduce the I3C master trait

rusty1968 committed:
Try unifying I2C and I3C errors.

rusty1968 committed:
Refactor i3c trait.

rusty1968 committed:
Mock an I3C concrete type that implements i2c and i3c behaviors.

rusty1968 committed:
Add sample code that shows how to use the i2c target

rusty1968 committed:
Event handler for i3c.

rusty1968 committed:
- cargo format
- client / server abstractions.

rusty1968 committed:
Make I2Ctarget more composable.

rusty1968 committed:
Update rustdoc for the fully featured I2Target convenience trait.

rusty1968 committed:
Refactor i3_target traits for better composability.

rusty1968 committed:
Update System Control related traits. Clock and Reset control.

rusty1968 committed:
Created a design section to the system control traits

rusty1968 committed:
Add rustdoc for the RSA trait

rusty1968 committed:
Adjust the definition of Digest according to Chris Frantz recommendations.

rusty1968 committed:
Revising ECDSA traits to interoperate with Digest.

rusty1968 committed:
Provide  a software implementation of the ECDSA trait.

rusty1968 committed:
Abstraction for OTP.

rusty1968 committed:
Revise symmetric cipher, mac.

rusty1968 committed:
Revise the BlockDevice traits.

rusty1968 committed:
Revise system control trait.

rusty1968 committed:
Revise Rsa traits.

rusty1968 committed:
Revise ecdsa traits : add Curve trait.

rusty1968 committed:
Software implementation of the Mac traits.

rusty1968 committed:
Software implementation of the mac traits.

rusty1968 committed:
Revise ecdsa traits to improve type safety and interoperability.

rusty1968 committed:
Add trait bounds to DigestOutput.

rusty1968 committed:
Revise EcdsaKeyGen so it returns a tuple (Priv, Pub) key pair to make it easier for the caller to use.

rusty1968 committed:
First draft of ASPEED I2C Driver Design Document.

rusty1968 committed:
Transfer Engine orchestration doc.

rusty1968 committed:
- ASPEED  I2C Driver design document

rusty1968 committed:
Removed all Send+Sync super-traits from the codebase. Updated documentation  accordingly.

rusty1968 committed:
Design document for proposed traits.

rusty1968 committed:
feat: Add composable OTP traits with soak programming support

- Add comprehensive error handling with ErrorKind enum covering all OTP operations
- Implement ErrorType trait pattern for consistent error handling across implementations
- Add OtpSession trait for session-based access control and resource management
- Add OtpRegions trait for multi-region OTP devices with per-region operations
- Add OtpProtection trait for region and global memory protection mechanisms
- Add OtpWriteTracking trait for devices with limited write attempts per location
- Add OtpVerification trait for programming verification and program-and-verify operations
- Add OtpIdentification trait for chip version detection and feature support queries
- Add OtpBulkOperations trait for efficient bulk read/write operations
- Add OtpMemoryLayout trait for detailed memory organization information
- Add OtpMultiWidth trait for devices supporting multiple data word sizes
- Add OtpSoakProgramming trait for extended timing programming of difficult bits
- Include comprehensive examples and test implementations demonstrating trait composition
- All traits follow consistent ErrorType pattern for type-safe error handling
- Traits are designed to be composable, allowing implementations to mix and match featuresSolve lifetime issues

rusty1968 committed:
refactor: Move ASPEED OTP to examples.

- Move otp_aspeed.rs from src/ to examples/ directory to reflect its role as an example implementation
- Remove otp_aspeed module from lib.rs core modules list
- Update imports in otp_aspeed.rs to reference proposed_traits crate instead of internal crate paths
- Add ExampleOtpController outside test module for use in main function
- Add comprehensive main() function demonstrating all ASPEED OTP capabilities:
  - Session management with chip detection
  - Multi-region operations (data, config, strap)
  - Strap bit programming with status tracking
  - Soak programming mode control
  - Protection mechanisms and status queries
  - Health checking and error handling
- Configure as proper Cargo example in Cargo.toml
- Separate example implementation from test-only MockOtpController
- Document how ASPEED-specific features extend generic OTP traits
- Maintain full trait compatibility

rusty1968 committed:
Dedicated OTP trait design.

rusty1968 committed:
Include information on Application Layer trait usage.

rusty1968 committed:
Add object-safe traits to support runtime selection of algos.

rusty1968 committed:
Revise object-safe trait for DigesOp.

rusty1968 committed:
Remove non-trait related stuff.

rusty1968 committed:
Example of an adapter between ASPEED trait and BlockDevice trait.

rusty1968 committed:
Add software digest trait implementations using RustCrypto

- Add comprehensive software_digest_complete.rs with static/dynamic APIs
- Add clean software_digest_impl.rs with essential functionality
- Implement SHA-256/384/512 using sha2 crate
- Fix borrow checker issues in test iteration
- Add complete test coverage and example usage
- Support both trait-based and direct digest computation

rusty1968 committed:
Implement comprehensive ECDSA reference using RustCrypto

- Added complete P256/ECDSA implementation using p256 and ecdsa crates
- Implemented all required trait signatures from proposed-traits/ecdsa.rs
- Added proper error handling with both ECDSA and common ErrorType traits
- Implemented endianness-aware ToBytes/FromBytes for all key/signature types
- Added key generation, signing, verification, and serialization support
- Fixed trait bound issues and API compatibility with elliptic curve crates
- Includes comprehensive tests covering all functionality
- All tests pass with working key generation, sign/verify, and serialization

rusty1968 committed:
refactor(digest): Replace Vec<u8> with static arrays for embedded-friendly digest outputs

Replace dynamic heap-allocated Vec<u8> outputs with compile-time fixed-size arrays
for zero-allocation digest operations in embedded environments.

Key changes:
- DigestAlgorithm::DigestOutput now uses [u8; N] instead of Vec<u8>
  - SHA-256: [u8; 32], SHA-384: [u8; 48], SHA-512: [u8; 64]
- Removed generic StaticDigestOp<D> in favor of algorithm-specific types:
  - Sha256StaticOp, Sha384StaticOp, Sha512StaticOp
- Updated static API methods to return fixed-size arrays
- Maintained Vec<u8> for dynamic API to preserve runtime flexibility
- All tests pass with improved type safety and zero heap allocation

Benefits:
- Zero heap allocation for static digest operations
- Better embedded systems support (no_std compatible)
- Compile-time buffer size validation
- Improved performance through stack allocation
- Clear separation between static (embedded) and dynamic (flexible) APIs

The implementation now provides true zero-allocation digest computation
for embedded systems while maintaining the dynamic API for applications
requiring runtime algorithm selection.

rusty1968 committed:
feat: Add static buffer management for zero-allocation dynamic digest

- Add compute_digest_static_buffer() method for runtime algorithm selection without heap allocation
- Add compute_digest_streaming_static_buffer() for streaming with static buffers
- Uses stack-allocated [u8; 64] worst-case buffer for SHA-512 compatibility
- Returns (buffer, actual_size) tuple for flexible embedded usage
- Provides runtime algorithm flexibility with static allocation guarantees
- Added comprehensive tests for static buffer functionality
- Updated main example to demonstrate static buffer benefits
- Perfect for embedded systems needing runtime hash selection without heap

rusty1968 committed:
feat: Add true interoperability between static and dynamic dispatch

- Implement DynamicDigestOp for all static operation types (Sha256StaticOp, Sha384StaticOp, Sha512StaticOp)
- Add create_digest_from_static() method to use static implementations through dynamic interface
- Add compute_digest_interop() methods demonstrating static dispatch through dynamic API
- Static operations can now be used as trait objects while maintaining type safety
- Added comprehensive tests showing interoperability between dispatch schemes
- Updated main example to demonstrate static operations used as dynamic trait objects
- Enables flexible architecture where static and dynamic implementations are interchangeable
- Perfect for systems needing both compile-time guarantees and runtime flexibility

rusty1968 committed:
Enum based polimorphism document.

rusty1968 committed:
Executable example of Enum-based polymorphism example using digest traits

rusty1968 committed:
Rename hash_mac_enum_based_polimorphism to hash_mac_enum_based_polimorphism.md

stevenlee7189 committed:
feat(i2c): add transaction direction to on_transaction_start for I2C target

Replace the old on_transaction_start(repeated: bool) method with a new version
that also includes transaction direction via an enum.

- Add `TransactionDirection` enum with `Read` and `Write` variants
- Update `I2CCoreTarget` trait to use `fn on_transaction_start(direction, repeated)`

Signed-off-by: Steven Lee <steven_lee@aspeedtech.com>

rusty1968 committed:
Merge pull request #7 from stevenlee7189/caliptra-test

feat(i2c): add transaction direction to on_transaction_start for I2C target

stevenlee7189 committed:
i2c: add return value to on_transaction_start

The I2CCoreTarget trait's on_transaction_start method now returns Result<Option<u8>, Error>.
This allows targets to:
- Return an initial byte immediately for read transactions
- Indicate setup errors

rusty1968 committed:
Merge pull request #8 from stevenlee7189/caliptra-test

i2c: add return value to on_transaction_start

rusty1968 committed:
Create aspeed-trng.txt
